addons/mail/models/statusbar_time_mixin.py

```diff
@@ -0,0 +1,76 @@
+import logging
+from datetime import timedelta
+
+from odoo import fields, models
+
+_logger = logging.getLogger(__name__)
+
+
+class StatusbarTimeMixin(models.AbstractModel):
+    _name = "statusbar.time.mixin"
+    _description = "Mixin to compute the time a record spent in each stage"
+    _stage_field = "stage_id"
+
+    json_status_time = fields.Json(
+        string="Status time", compute="_compute_json_status_time", help="JSON that maps stage ids to seconds spent")
+
+    def _compute_json_status_time(self):
+        """
+        Computes json_status_time, a Json field stored as { <stage_id (str)>: <time_spent_in_seconds (int)> }
+
+            e.g. {"1": 1230, "3": 2220, "5": 14}
+        """
+
+        field = self.env['ir.model.fields'].sudo().search_fetch([
+            ('model', '=', self._name),
+            ('name', '=', self._stage_field),
+        ], ['id', 'tracking'], limit=1)
+        if (field.tracking == 0):
+            self.json_status_time = False
+            _logger.warning(
+                'Field "%s" on model "%s" must have tracking=True for the computation of statusbar time.',
+                self._stage_field, self._name
+            )
+            return
+
+        query = """
+               SELECT m.res_id,
```

Comment:
Currently we have an index on field and mail_message_id on mail.tracking.value so seems ok. We also have a combined index on model / res_id for mail.message. However I don't know how this query will behave as a whole. Would be curious to see how this query behaves on our prod... did you try ? If not, could be interesting to get in touch with our internal team and see a bit what is the execution plan :)